File: usermodel.py
from mysqlmodel import MysqlModel
from flask import current_app, abort
from redisstring import RedisString

import util, time


class UserModel(MysqlModel):
	colmap = ['id', 'username', 'password']

	def __init__(self):
		self.token = None

	def update_token(self):
		token = util.gen_random_string()
		self.token = token
		RedisString("token2userid_" + token).set(self.id)
		return token

	@classmethod
	def find_userid_by_token(cls, token):
		userid = RedisString("token2userid_" + token).get()
		if userid is None:
			return False
		return userid

	@classmethod
	def init_data_structure(cls):
		current_app.username2userid = dict()
		current_app.username2user = dict()

	def after_parse(self):
		current_app.username2userid[self.username] = self.id
		current_app.username2user[self.username] = self

	@classmethod
	def login(cls, username, password):
		user = current_app.username2user.get(username, None)
		if (user is None) or user.password != password: return False
		user.update_token()
		return user

File: src/eleme/user_model.go
package eleme

import "strconv"

type UserModel struct {
	MysqlModel
	realid int
	realidString string
	username string
	password string
	token string
}

var username2user = map[string] *UserModel {}
var token2user = map[string] *UserModel {}

func (this *UserModel) fetch(id string) *UserModel {
	ret := this.MysqlModel.fetch(id)
	if ret == nil { return nil }
	return ret.(*UserModel)
}

func (this *UserModel) create(id int, username, password string) *UserModel {
	ret := new(UserModel)
	ret.realidString = strconv.Itoa(id)
	ret.id = "User_" + ret.realidString
	ret.realid = id
	ret.username = username
	ret.password = password
	ret.token = md5_hex(password)

	datapool[ret.id] = ret
	username2user[username] = ret
	token2user[ret.token] = ret

	return ret
}

func (this *UserModel) findUserIdByToken(token string) string {
	user, ok := token2user[token]
	if !ok { return "" }
	return user.id
}

func (this *UserModel) login(username, password string) *UserModel {
	user, ok := username2user[username]
	if !ok { return nil }
	if user.password != password { return nil }
	return user
}

var userModel *UserModel

